apollo-federation/src/sources/graphql/federated_query_graph/builder.rs

```diff
@@ -9,9 +32,183 @@ pub(crate) struct FederatedQueryGraphBuilder;
+            let node = super::AbstractNode {
+                subgraph_type,
+                provides_directive: None,
+            };
+            (ty, builder.add_abstract_node(type_name, node.into())?)
+        }
+    };
+    existing_types.insert(output_type_definition_position, index);
+    match ty {
+        Some(Either::Left(concrete_ty)) => concrete_ty
+            .field_positions(subgraph_schema.schema())?
+            .filter(|field| is_field_ignored(field.clone().into(), subgraph_schema))
+            .try_for_each(|subgraph_field| {
+                subgraph_schema
+                    .get_type(subgraph_field.type_name.clone())
+                    .and_then(OutputTypeDefinitionPosition::try_from)
+                    .and_then(|ty| process_type(ty, subgraph_schema, builder, existing_types))
+                    .and_then(|next_index| {
+                        builder
+                            .add_concrete_field_edge(
+                                index,
+                                next_index,
+                                subgraph_field.field_name.clone(),
+                                IndexSet::new(),
+                                ConcreteFieldEdge {
+                                    subgraph_field,
+                                    requires_condition: None,
+                                }
+                                .into(),
+                            )
+                            .map(drop)
+                    })
+            })?,
+        Some(Either::Right(abstract_ty)) => subgraph_schema
+            .possible_runtime_types(abstract_ty.into())?
+            .into_iter()
+            .try_for_each(|ty| {
+                process_type(ty.clone().into(), subgraph_schema, builder, existing_types).and_then(
+                    |next_index| {
+                        builder
+                            .add_type_condition_edge(
+                                index,
+                                next_index,
+                                TypeConditionEdge::new(ty.into()).into(),
+                            )
+                            .map(drop)
+                    },
+                )
+            })?,
+        None => {}
+    }
+    Ok(index)
+}
+
+fn is_field_ignored(
+    field_definition_position: FieldDefinitionPosition,
+    subgraph_schema: &ValidFederationSchema,
+) -> bool {
+    if field_definition_position.field_name().starts_with("__") {
+        return true;
+    }
+
+    if let FieldDefinitionPosition::Object(position) = &field_definition_position {
+        let query_def = ObjectTypeDefinitionPosition {
+            type_name: name!("Query"),
+        };
+        let entities = query_def.field(name!("_entities"));
+        let service = query_def.field(name!("_service"));
+        if [entities, service].contains(position) {
+            return true;
+        }
     }
+
+    subgraph_schema
+        .subgraph_metadata()
+        .unwrap()
```

Comment:
Router folks have told me we want to try to avoid calling code that can panic. I think you can avoid the `.unwrap()` by:
1. Changing `is_field_ignored()` to return a `Result<bool, FederationError>`.
2. Use e.g. `let ... else { return Err(FederationError::internal()) }` here to get metadata.
3. Changing the callsites to use `.process_results()` from `itertools`. E.g. the `filter()` call above becomes
```rust
.map(|field| is_field_ignored(field.clone().into(), subgraph_schema).map(|is_ignored| (!is_ignored).then_some(field)))
.process_results(|iter| {
    iter.flatten().try_for_each(|subgraph_field| {
        // ...
    })
})??,
```